src/main/java/org/candlepin/subscriptions/product/OfferingSyncController.java

```diff
@@ -196,4 +206,87 @@ public Stream<String> saveOfferings(
   public void deleteOffering(String sku) {
     offeringRepository.deleteById(sku);
   }
+
+  /**
+   * Sync offering state based on a UMB message.
+   *
+   * <p>See syncRootSku and syncChildSku for more details.
+   *
+   * @param productXml UMB message for product
+   * @return result describing results of sync (for testing purposes)
+   */
+  @javax.transaction.Transactional
+  public SyncResult syncUmbProductFromXml(String productXml) throws JsonProcessingException {
+    return syncUmbProduct(
+        umbMessageMapper
+            .readValue(productXml, org.candlepin.subscriptions.umb.CanonicalMessage.class)
+            .getPayload()
+            .getSync()
+            .getOperationalProduct());
+  }
+
+  private SyncResult syncUmbProduct(UmbOperationalProduct umbOperationalProduct) {
+    if (umbOperationalProduct.getSku().startsWith("SVC")) {
+      syncChildSku(umbOperationalProduct.getSku());
+      return SyncResult.FETCHED_AND_SYNCED;
+    } else {
+      SyncResult result = syncRootSku(umbOperationalProduct);
+      if (result == SyncResult.FETCHED_AND_SYNCED) {
+        // we must assume that any SKU we get a message for may be a derived SKU,
+        // but we'll check our cache of product data and only actually operate on offerings having
+        // the SKU as a derived SKU
+        syncDerivedSku(umbOperationalProduct.getSku());
+      }
+      return result;
+    }
+  }
+
+  /**
+   * Sync the offering state using only the UMB message if possible, otherwise sync the offering
+   * from the RHIT product service.
+   *
+   * @param umbOperationalProduct product definition from a UMB message
+   * @see UpstreamProductData#offeringFromUmbData
+   */
+  private SyncResult syncRootSku(UmbOperationalProduct umbOperationalProduct) {
+    Optional<Offering> existing = offeringRepository.findById(umbOperationalProduct.getSku());
+    Optional<Offering> newState =
+        UpstreamProductData.offeringFromUmbData(
+            umbOperationalProduct, existing.orElse(null), productService);
+    if (newState.isPresent()) {
+      return syncOffering(newState.get(), existing);
+    } else {
+      LOGGER.warn(
+          "Unable to sync offering from UMB message for sku={}, because product service has no records for it",
+          umbOperationalProduct.getSku());
+      return SyncResult.SKIPPED_NOT_FOUND;
+    }
+  }
+
+  /**
+   * Sync all offerings affected by a change in the child SKU.
+   *
+   * <p>(Future work could reduce the need to sync affected offerings by caching child SKU eng IDs
+   * and only queueing offering syncs when there are changes).
+   *
+   * @param sku child SKU (starts with "SVC" by convention)
+   */
+  private void syncChildSku(String sku) {
+    Set<String> parentSkus =
+        offeringRepository.findSkusForChildSku(sku).collect(Collectors.toSet());
+    parentSkus.forEach(this::enqueueOfferingSyncTask);
+    offeringRepository.findSkusForDerivedSkus(parentSkus).forEach(this::enqueueOfferingSyncTask);
```

Comment:
:+1:

Reply:
Actually, I realized why I wrote it this way...  Calling `syncDerivedSku` in a loop would cause a separate query for each matching parent SKU. I'm leaving it as-is and leaving some comments to explain this.